qiskit/quantum_info/states/stabilizerstate.py

```diff
@@ -0,0 +1,591 @@
+        """
+        # Resetting all qubits does not require sampling or RNG
+        if qargs is None:
+            return StabilizerState(Clifford((np.eye(2 * self.data.num_qubits))))
+
+        for qubit in qargs:
+            # Apply measurement and get classical outcome
+            outcome = self._measure_and_update(qubit, 0)
+
+            # Use the outcome to apply X gate to any qubits left in the
+            # |1> state after measure, then discard outcome.
+            if outcome == 1:
+                _append_x(self.data, qubit)
+
+        return self
+
+    def measure(self, qargs=None):
+        """Measure subsystems and return outcome and post-measure state.
+
+        Note that this function uses the QuantumStates internal random
+        number generator for sampling the measurement outcome. The RNG
+        seed can be set using the :meth:`seed` method.
+
+        Args:
+            qargs (list or None): subsystems to sample measurements for,
+                                  if None sample measurement of all
+                                  subsystems (Default: None).
+
+        Returns:
+            tuple: the pair ``(outcome, state)`` where ``outcome`` is the
+                   measurement outcome string label, and ``state`` is the
+                   collapsed post-measurement stabilizer state for the
+                   corresponding outcome.
+        """
+        if qargs is None:
+            qargs = range(self.data.num_qubits)
+
+        outcome = ''
+        for qubit in qargs:
+            randbit = self._rng.randint(2)
+            outcome = str(self._measure_and_update(qubit, randbit)) + outcome
+        return outcome, self
+
+    def sample_memory(self, shots, qargs=None):
+        """Sample a list of qubit measurement outcomes in the computational basis.
+
+        Args:
+            shots (int): number of samples to generate.
+            qargs (None or list): subsystems to sample measurements for,
+                                if None sample measurement of all
+                                subsystems (Default: None).
+
+        Returns:
+            np.array: list of sampled counts if the order sampled.
+
+        Additional Information:
+
+            This function implements the measurement :meth:`measure` method.
+
+            The seed for random number generator used for sampling can be
+            set to a fixed value by using the stats :meth:`seed` method.
+        """
+        memory = []
+        for _ in range(shots):
+            # copy the StabilizerState since measure updates it
+            stab = copy.deepcopy(self)
+            memory.append(stab.measure(qargs)[0])
+        return memory
+
+    # -----------------------------------------------------------------------
+    # Helper functions for calculating the measurement
+    # -----------------------------------------------------------------------
+    def _measure_and_update(self, qubit, randbit):
```

Comment:
I think the measurement/reset issues could be fixed by modifying this function to do its own random number generatio and return the updated state. Essentially this is a single-qubit measure function. Something like:

```python
def _measure_qubit(self, qubit):
    """Returns: (outcome, collapsed_state)"""
    z_anticommuting = np.any(self.data.stabilizer.X[:, qubit])
    
    if z_anticommuting == 0:    
        # Deterministic outcome - measuring it will not change the StabilizerState
        aux_pauli = Pauli(num_qubits * 'I')
        for i in range(num_qubits):
            if self.data.table.X[i][qubit]:
                aux_pauli = self._rowsum_deterministic(aux_pauli, i + num_qubits)
        outcome = aux_pauli.phase
        return outcome, self
    else:
        # Non-deterministic outcome
        outcome = self._rng.integers(2)
        ret = self.copy() # this copies current state of rng generator too
        table = ret.data.table
        # code to update table ...
        return outcome, ret
```

Reply:
I've slightly updated the function, but didn't make the following changes:
- We still need `randbit` as an input for the recursive function `_get_probablities` (see last 2 lines of the code)
- The function is still called `_measure_and_update`, returns `outcome` and updates the state. 
Otherwise, whenever someone calls the above suggested function, which returns 2 items, it will be:
`_measure_qubit(self, qubit)[0]`, which might be confusing.